File: leetcode/problems/search_a_2d_matrix/solution.cpp
class Solution {
public:
    bool searchMatrix(vector<vector<int>>& matrix, int target) {
        auto cmp = [](int a, const vector<int>& b){
            return a < b[0];
        };
        auto it = upper_bound(matrix.begin(), matrix.end(), target, cmp);
        if(it == matrix.begin()) {
            return false;
        }
        it--;
        return binary_search(it->begin(), it->end(), target);
    }
};

File: leetcode/problems/search_a_2d_matrix/solution.rs
use std::cmp::Ordering;
impl Solution {
    pub fn search_matrix(matrix: Vec<Vec<i32>>, target: i32) -> bool {
        let r = matrix.binary_search_by(|row| {
            if row[0] <= target {
                Ordering::Less
            } else {
                Ordering::Greater
            }
        })
        .unwrap_err();
        r > 0 && matrix[r-1].binary_search(&target).is_ok()
    }
}